src/slot_provider/clock.rs

```diff
@@ -1,88 +1,184 @@
 //! Clock based SlotProvider
-//! When a fn does not finish within a slot (12 seconds), the next call is made concurrently, up to a buffer limit. When the buffer limit is reached, the passed fn will not be called for a new slot, but wait in a FIFO queue until a previous slot has finished.
-use ethers::providers::StreamExt;
-use eyre::Result;
-use futures::Future;
-use tokio::time::{interval, timeout, Duration};
-
-use crate::slot_provider::{
-    wait_until_slot_start, Slot, SlotProvider, GENESIS_SLOT_TIME, SLOT_PERIOD_SECONDS,
-};
+use async_trait::async_trait;
+use chrono::Utc;
+use futures::stream::StreamExt;
+use std::collections::VecDeque;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::Arc;
+use tokio::sync::{mpsc, Mutex};
+use tokio::time::{interval, Duration};
 use tokio_stream::wrappers::IntervalStream;
 
+use super::{Slot, SlotProvider};
+
+const SLOT_PERIOD_SECONDS: u32 = 12;
+const MAX_SLOTS: usize = 32;
+
+/// Waits until the start of the next slot.
+/// Much of our code depends on what the current slot is, and wants to answer as fast as possible,
+/// therefore we sometimes want to align our code with the start of the slot.
+async fn wait_until_next_slot() {
+    let current_slot = Slot::from_timestamp(Utc::now().timestamp() as u64).unwrap();
+    let next_slot = Slot {
+        number: current_slot.number + 1,
+    };
+    let next_slot_start = next_slot.to_date_time();
+    // NOTE: doesn't account for leap seconds.
+    let seconds_until_next_slot = next_slot_start.timestamp() - Utc::now().timestamp();
+
+    tokio::time::sleep(tokio::time::Duration::from_secs(
+        seconds_until_next_slot as u64,
+    ))
+    .await;
+}
+
 pub struct SystemClockSlotProvider {
-    num_slots: Option<usize>,
+    counter: AtomicUsize,
+    max_count: Option<usize>,
+    slots: Mutex<VecDeque<Slot>>,
 }
 
 impl SystemClockSlotProvider {
-    pub fn new() -> Self {
-        Self { num_slots: None }
+    fn internal_new(notify: mpsc::Sender<()>, max_count: Option<usize>) -> Arc<Self> {
```

Comment:
Not sure about this internal_new pattern. What do you think? It's designed so:
* the "normal use" `new` which does not have a limit does not worry about one.
* the test use `new` which has a limit still reuses all the code from the normal new.

But then, how often will we call this. Maybe twice.. thinking about it more just taking limit as the second argument with Option is probably better. Oh hah, almost forgot, or the way you had it, using a setter. Mm, I'd prefer instead of modifying internal behavior by a set which can be executed at any point. Is it variant A or B for the lifetime of this var. But again, all nitpicking, almost never gets called, don't really care.

Reply:
I like it generally. Don't care about it too much either.

I think new returning an `Arc<Self>` instead of `Self` would be a bit surprising to me. 
Is there any specific reason why you moved the "wrapping" into an `Arc` inside these functions instead of doing it on the callers side ?

Reply:
There's a clone inside of it to then hand off one reference to main, and one to a thread that gets spawned. Can't do that without Arc although I'm sure there's other ways to both be able to have a reference in main that gets you the latest slot and spawn a thread which will continually update the scheduler a new slot is available.

Reply:
Got it, thanks for the explanation. Looks fine to me